src/main/java/org/quiltmc/quiltmappings/loom/QuiltMappingsOnLoomPlugin.java

```diff
@@ -155,30 +182,46 @@ private void extractFile(File dependency, File output, Predicate<File> filter) {
-        //     File definitions = project.file(".gradle/qm/unpick_definitions_" + minecraftVersion + quiltMappingsBuild + ".unpick");
-        //     File metadata = project.file(".gradle/qm/unpick_metadata_" + minecraftVersion + quiltMappingsBuild + ".json");
-        //
-        //     if (!definitions.exists() || !metadata.exists()) {
-        //         List<File> quiltMappings = new ArrayList<>(project.getConfigurations().detachedConfiguration(project.getDependencies().create(this.quiltMappings)).resolve());
-        //
-        //         extractFile(quiltMappings.get(0), definitions, file -> file.getName().endsWith("definitions.unpick"));
-        //         extractFile(quiltMappings.get(0), metadata, file -> file.getName().endsWith("unpick.json"));
-        //     }
-        //
-        //     String configuration = Constants.Configurations.MAPPING_CONSTANTS;
-        //     if (project.getConfigurations().getByName(configuration).getDependencies().isEmpty()) {
-        //         String dependency = quiltMappings.substring(0, quiltMappings.lastIndexOf(":")) + ":constants";
-        //         project.getDependencies().add(configuration, dependency);
-        //     }
-        //
-        //     return UnpickData.read(metadata.toPath(), definitions.toPath());
-        // }
+        @Override
+        public UnpickLayer.UnpickData getUnpickData() throws IOException {
+            Project project = layer.project;
+            String quiltMappings = layer.quiltMappings;
+            String minecraftVersion = layer.context.minecraftProvider().minecraftVersion();
+
+            String quiltMappingsBuild = "+build" + quiltMappings.substring(quiltMappings.lastIndexOf("."), quiltMappings.lastIndexOf(":"));
+            File definitions = project.file(".gradle/qm/unpick_definitions_" + minecraftVersion + quiltMappingsBuild + ".unpick");
+            File metadata = project.file(".gradle/qm/unpick_metadata_" + minecraftVersion + quiltMappingsBuild + ".json");
+
+            if (!definitions.exists() || !metadata.exists()) {
+                List<File> mappingsList = new ArrayList<>(project.getConfigurations().detachedConfiguration(project.getDependencies().create(quiltMappings)).resolve());
+
+                layer.extractFile(mappingsList.get(0), definitions, file -> file.getName().endsWith("definitions.unpick"));
+                layer.extractFile(mappingsList.get(0), metadata, file -> file.getName().endsWith("unpick.json"));
+            }
+
+            String configuration = Constants.Configurations.MAPPING_CONSTANTS;
+            if (project.getConfigurations().getByName(configuration).getDependencies().isEmpty()) {
+                String dependency = quiltMappings.substring(0, quiltMappings.lastIndexOf(":")) + ":constants";
+                project.getDependencies().add(configuration, dependency);
+            }
```

Comment:
Apparently this only runs when the layer is first generated, so it should be moved to another method that always runs

Reply:
Also, the constants jar has a dependency on hashed, which makes the `unpickJar` task fail because the configuration resolves both the constants and hashed jars, so hashed should be excluded (like how you would exclude in a buildscript with `implemetation("com.example:lib") { exclude group: "com.example" module: "dependency" }`)

Reply:
And placing the code in `QuiltMappingsOnLoomExtension#mappings` makes it work everytime